### PR TITLE
Fix ASGI instrumentation default span name

### LINKED ISSUE #146: Fix asgi instrumentation default span name
Similar to [wsgi](https://github.com/open-telemetry/opentelemetry-python/blob/master/instrumentation/opentelemetry-instrumentation-wsgi/src/opentelemetry/instrumentation/wsgi/__init__.py#L171), the default span name for SERVER spans should be `HTTP <method>` as outlined by the [specs](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md#name).

Part of https://github.com/open-telemetry/opentelemetry-python-contrib/issues/434

### PR DESCRIPTION
**Pull request by adamantike**
# Description

Fixes ASGI default span name for SERVER spans, to be `HTTP <method>`.

Fixes #146.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Updated unit tests, executed with `tox -e test-instrumentation-asgi`.

# Does This PR Require a Core Repo Change?

- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 1]
```diff
@@ -144,12 +144,12 @@ def get_default_span_details(scope: dict) -> Tuple[str, dict]:
         scope: the asgi scope dictionary
 
     Returns:
-        a tuple of the span, and any attributes to attach to the
-        span.
+        a tuple of the span name, and any attributes to attach to the span.
     """
     method_or_path = scope.get("method") or scope.get("path")
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:149]: I believe the logic should be: Use the path if it exists. If not, use "HTTP <method>".

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 2]
```diff
@@ -194,7 +194,7 @@ async def __call__(self, scope, receive, send):
 
         try:
             with self.tracer.start_as_current_span(
-                span_name + " asgi", kind=trace.SpanKind.SERVER,
+                span_name, kind=trace.SpanKind.SERVER,
             ) as span:
                 if span.is_recording():
                     attributes = collect_request_attributes(scope)
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:200]: I think the async methods need to be fixed too.

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py (hunk 1 above)
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:149]: I understood, based on the current [wsgi implementation](https://github.com/adamantike/opentelemetry-python-contrib/blob/60222c2282914a484f4c6f41ff1018e09e931f9e/instrumentation/opentelemetry-instrumentation-wsgi/src/opentelemetry/instrumentation/wsgi/__init__.py#L177-L179), and the [span name convention](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md#name), that we would always want the span name to be `HTTP {METHOD|PATH}`.

Do you consider that we should prioritize `{PATH}` over `HTTP {METHOD}`, when the path is available?

Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:149]: Yeah `HTTP METHOD` is what we default to when there is no route found. Take a look at [django](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py#L69) and [flask](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py#L76).

Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py]: Makes sense, thanks for the heads up! Comments applied.

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 3]
```diff
@@ -205,7 +206,7 @@ async def __call__(self, scope, receive, send):
                 @wraps(receive)
                 async def wrapped_receive():
                     with self.tracer.start_as_current_span(
-                        span_name + " asgi." + scope["type"] + ".receive"
+                        span_name + " " + scope["type"] + ".receive"
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:209]: `scope["type"].receive` is not defined in the specs. I am wondering what we should do with these spans.
@codeboten thoughts?

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 4]
```diff
@@ -144,12 +144,13 @@ def get_default_span_details(scope: dict) -> Tuple[str, dict]:
         scope: the asgi scope dictionary
 
     Returns:
-        a tuple of the span, and any attributes to attach to the
-        span.
+        a tuple of the span name, and any attributes to attach to the span.
     """
-    method_or_path = scope.get("method") or scope.get("path")
+    span_name = scope.get("path") or "HTTP {}".format(
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:149]: We should `strip()` the path as well.

**Comment by lzchen** [comment]
@adamantike 
Any updates on this PR?

**Comment by adamantike** [comment]
@lzchen, sorry for the delay, I was waiting for a follow-up regarding the comment on `scope["type"].receive`.

I have rebased, and applied your other comment.

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py (hunk 3 above)
Comment (codeboten) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:220]: im wondering if the SpanKind should be set here in addition to having the name of the span reflect the sender/receiver as per the messaging semantic conventions: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#span-name

**Comment by codeboten** [comment]
@adamantike apologies for the delay, i completely missed this in my notifications.

**Comment by codeboten** [comment]
@adamantike have you had a chance to look at @lzchen's comment?

**Comment by adamantike** [comment]
@lzchen @codeboten, I have replaced the separator for destination and operation names. Please, let me know if that's what we expect based on the specs.

**Comment by adamantike** [comment]
Rebased and fixed conflicts!

**Comment by lzchen** [comment]
@adamantike 

Please fix the lint issues, then we can merge this :)

**Comment by adamantike** [comment]
It seems the CI error on the `lint` stage is failing globally for all PRs? It seems `opentelemetry-test` has been removed from PyPI:

```
lint run-test-pre: commands[37] | python -m pip install -e '/home/runner/work/opentelemetry-python-contrib/opentelemetry-python-contrib/instrumentation/opentelemetry-instrumentation-tornado[test]'
Obtaining file:///home/runner/work/opentelemetry-python-contrib/opentelemetry-python-contrib/instrumentation/opentelemetry-instrumentation-tornado
Collecting tornado>=6.0
  Downloading tornado-6.1-cp38-cp38-manylinux2010_x86_64.whl (427 kB)
Collecting opentelemetry-instrumentation==0.21b0
  Downloading opentelemetry_instrumentation-0.21b0-py3-none-any.whl (16 kB)
Collecting opentelemetry-api==1.2.0
  Downloading opentelemetry_api-1.2.0-py3-none-any.whl (37 kB)
Collecting opentelemetry-semantic-conventions==0.21b0
  Downloading opentelemetry_semantic_conventions-0.21b0-py3-none-any.whl (20 kB)
Collecting opentelemetry-util-http==0.21b0
  Downloading opentelemetry_util_http-0.21b0-py3-none-any.whl (3.6 kB)
ERROR: Could not find a version that satisfies the requirement opentelemetry-test==0.21b0 (from opentelemetry-instrumentation-tornado[test]) (from versions: none)
ERROR: No matching distribution found for opentelemetry-test==0.21b0
ERROR: InvocationError for command /home/runner/work/opentelemetry-python-contrib/opentelemetry-python-contrib/.tox/lint/bin/python -m pip install -e '/home/runner/work/opentelemetry-python-contrib/opentelemetry-python-contrib/instrumentation/opentelemetry-instrumentation-tornado[test]' (exited with code 1)
```

**Comment by lzchen** [comment]
Ahh I see, should be fixed after [this](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/510).

**Review by ocelotl — APPROVED** [review]

**Review by codeboten — APPROVED** [review]

**Pull request merged**